systemdns/dnsevtlog/events.go

```diff
@@ -0,0 +1,139 @@
+//go:build windows
+// +build windows
+
+package dnsevtlog
+
+// This code is copied from Promtail v1.6.2-0.20231004111112-07cbef92268a with minor changes.
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"golang.org/x/sys/windows"
+
+	"github.com/safing/portbase/log"
+	"github.com/safing/portmaster/systemdns/dnsevtlog/win_eventlog"
+)
+
+type Subscription struct {
+	subscription win_eventlog.EvtHandle
+	fetcher      *win_eventlog.EventFetcher
+
+	eventLogName  string
+	eventLogQuery string
+
+	ready bool
+	done  chan struct{}
+	wg    sync.WaitGroup
+	err   error
+}
+
+// NewSubscription create a new windows event subscriptions.
+func NewSubscription() (*Subscription, error) {
+	sigEvent, err := windows.CreateEvent(nil, 0, 0, nil)
+	if err != nil {
+		return nil, err
+	}
+	defer windows.CloseHandle(sigEvent)
+
+	t := &Subscription{
+		eventLogName:  "Microsoft-Windows-DNS-Client/Operational",
+		eventLogQuery: "*",
+		done:          make(chan struct{}),
+		fetcher:       win_eventlog.NewEventFetcher(),
+	}
+
+	subsHandle, err := win_eventlog.EvtSubscribe(t.eventLogName, t.eventLogQuery)
+	if err != nil {
+		return nil, fmt.Errorf("error subscribing to windows events: %w", err)
+	}
+	t.subscription = subsHandle
+
+	return t, nil
+}
+
+// loop fetches new events and send them to via the Loki client.
+func (t *Subscription) ReadWorker() {
+	t.ready = true
+	t.wg.Add(1)
+	interval := time.NewTicker(time.Second)
+	defer func() {
+		t.ready = false
+		t.wg.Done()
+		interval.Stop()
+	}()
```

Comment:
I guess `ReadWorker` is expected to run in a goroutine. so access to `read` is unprotected if it's launched multiple times.
Also, `wg.Add(1)` should be called before the go-routine is launched since there could be a race condition when starting the `ReadWorker` and immediatley calling `Stop` which could cause a panic. See https://pkg.go.dev/sync#WaitGroup.Add.

Better add a dedicated `Start()` routine that adds to the wg and kicks of the goroutine